<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Quickstart &mdash; taxtastic 0.5.2 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.5.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="taxtastic 0.5.2 documentation" href="index.html" />
    <link rel="next" title="Manipulating the NCBI taxonomy" href="taxonomy.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="taxonomy.html" title="Manipulating the NCBI taxonomy"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="installation.html" title="Installation"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">taxtastic 0.5.2 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="quickstart">
<h1>Quickstart<a class="headerlink" href="#quickstart" title="Permalink to this headline">¶</a></h1>
<div class="section" id="minimal-example">
<h2>Minimal example<a class="headerlink" href="#minimal-example" title="Permalink to this headline">¶</a></h2>
<p>Say we have the following:</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">seqs.fasta</span></tt>: The multiply aligned reference sequences in FASTA format.</li>
<li><tt class="docutils literal"><span class="pre">tree.nwk</span></tt>: The tree built from reference alignment, in Newick format.</li>
<li><tt class="docutils literal"><span class="pre">tree_stats.txt</span></tt>: The log file from FastTree or the statistics file from RAxML/phyml.</li>
</ul>
<p>Build a reference package <tt class="docutils literal"><span class="pre">my.refpkg</span></tt> for a locus <tt class="docutils literal"><span class="pre">locus_name</span></tt> (e.g. <tt class="docutils literal"><span class="pre">16s_rRNA</span></tt>) in a single command as follows:</p>
<div class="highlight-python"><div class="highlight"><pre>taxit create -l 16s_rRNA -P my.refpkg \
    --aln-fasta seqs.fasta \
    --tree-stats tree_stats.txt \
    --tree-file tree.nwk
</pre></div>
</div>
</div>
<div class="section" id="taxonomically-equipped-reference-package-example">
<h2>Taxonomically equipped reference package example<a class="headerlink" href="#taxonomically-equipped-reference-package-example" title="Permalink to this headline">¶</a></h2>
<p>The fun really begins when we incorporate taxonomic information.
Say in addition to the above files you also have:</p>
<ul>
<li><p class="first"><tt class="docutils literal"><span class="pre">tax_ids.txt</span></tt>: A list of all the tax_ids that the sequences in your refpkg use, one per line in a text file.  Blank lines and Python style comments in this file will be ignored, and repeated entries are only considered once.</p>
</li>
<li><p class="first"><tt class="docutils literal"><span class="pre">seq_info.csv</span></tt>: A CSV file describing each of the aligned sequences.  It begins with one line giving the field names:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="s">&quot;seqname&quot;</span><span class="p">,</span><span class="s">&quot;accession&quot;</span><span class="p">,</span><span class="s">&quot;tax_id&quot;</span><span class="p">,</span><span class="s">&quot;species_name&quot;</span><span class="p">,</span><span class="s">&quot;is_type&quot;</span>
</pre></div>
</div>
<p>This is followed by one line for each sequence in your multiple alignment. Only <tt class="docutils literal"><span class="pre">seqname</span></tt> and <tt class="docutils literal"><span class="pre">tax_id</span></tt> are required. <tt class="docutils literal"><span class="pre">seqname</span></tt> should match the ID of the sequence in the FASTA file and must be unique. <tt class="docutils literal"><span class="pre">accession</span></tt> provides an additional identifier for the sequence, which can (but need not) be the same as <tt class="docutils literal"><span class="pre">seqname</span></tt>.  In our work, <tt class="docutils literal"><span class="pre">seqname</span></tt> is often an RDP accession number and <tt class="docutils literal"><span class="pre">accession</span></tt> is the NCBI accession number corresponding to that RDP entry (note that an NCBI accession number may not uniquely identify a sequence, as in tha case of a genomic sequence with multiple 16S rRNA gene loci).``tax_id`` is the entry in the taxonomy this sequence is mapped to, and <tt class="docutils literal"><span class="pre">species_name</span></tt> is the name associated with that entry. <tt class="docutils literal"><span class="pre">is_type</span></tt> indicates whether this sequence is from a type strain or not (should be <tt class="docutils literal"><span class="pre">TRUE</span></tt> or <tt class="docutils literal"><span class="pre">FALSE</span></tt>).  If you aren&#8217;t dealing with typestrains, just set it to <tt class="docutils literal"><span class="pre">FALSE</span></tt> for all your sequences.</p>
</li>
</ul>
<p>Now there are a few additional steps to construct a taxonomically-annotated refpkg.</p>
<p>First we have to represent the tax_ids as a taxonomy.  You will need a copy of the NCBI taxonomy in a format taxtastic understands in order to extract the minimal taxonomy containing the tax_ids in <tt class="docutils literal"><span class="pre">tax_ids.txt</span></tt>.  Run:</p>
<div class="highlight-python"><div class="highlight"><pre>taxit new_database -d taxonomy.db
</pre></div>
</div>
<p>This will download zipped files containing the NCBI taxonomy into the current directory and load the data into an sqlite databse named <tt class="docutils literal"><span class="pre">taxonomy.db</span></tt>.  This takes a while!  Plan to go get a cup of coffee.  Then run:</p>
<div class="highlight-python"><div class="highlight"><pre>taxit taxtable -d taxonomy.db -t tax_ids.txt -o taxa.csv
</pre></div>
</div>
<p>The output, <tt class="docutils literal"><span class="pre">taxa.csv</span></tt>, is a CSV file containing the minimum subtaxonomy of <tt class="docutils literal"><span class="pre">taxonomy.db</span></tt> which contains all the entries in <tt class="docutils literal"><span class="pre">tax_ids.txt</span></tt>.</p>
<p>Now we are ready to build the refpkg:</p>
<div class="highlight-python"><div class="highlight"><pre>taxit create -l locus_name -P my.refpkg \
    --taxonomy taxa.csv \
    --aln-fasta seqs.fasta \
    --seq-info seq_info.csv \
    --tree-stats tree_stats.txt \
    --tree-file tree.nwk
</pre></div>
</div>
<p>You may also want to add some metadata fields using the options <tt class="docutils literal"><span class="pre">--author</span></tt>, <tt class="docutils literal"><span class="pre">--description</span></tt>, and <tt class="docutils literal"><span class="pre">--package-version</span></tt>, but these are optional. In addition, you may package in profile information and a Stockholm alignment:</p>
<div class="highlight-python"><div class="highlight"><pre>--aln-sto seqs.sto \
--profile align_profile
</pre></div>
</div>
<p>where <tt class="docutils literal"><span class="pre">align_profile</span></tt> is a profile that can be used to align query sequences using HMMER3 or Infernal, and <tt class="docutils literal"><span class="pre">seqs.sto</span></tt> is the reference alignment in Stockholm format. This is handy if you plan to use <tt class="docutils literal"><span class="pre">cmalign</span></tt> (in the Infernal suite of tools) or <tt class="docutils literal"><span class="pre">hmmalign</span></tt> (in the HMMER3 suite of tools) to perform your alignments.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Quickstart</a><ul>
<li><a class="reference internal" href="#minimal-example">Minimal example</a></li>
<li><a class="reference internal" href="#taxonomically-equipped-reference-package-example">Taxonomically equipped reference package example</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="installation.html"
                        title="previous chapter">Installation</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="taxonomy.html"
                        title="next chapter">Manipulating the NCBI taxonomy</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/quickstart.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="taxonomy.html" title="Manipulating the NCBI taxonomy"
             >next</a> |</li>
        <li class="right" >
          <a href="installation.html" title="Installation"
             >previous</a> |</li>
        <li><a href="index.html">taxtastic 0.5.2 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2011-2013, Noah Hoffman, Erick Matsen, Brian Hodges.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.1.
    </div>
  </body>
</html>